File: functions/src/scoring.js
/**
 * Scoring Cloud Functions.
 *
 * Calculates game results, stone exchange, and match progression.
 *
 * Scoring rules:
 * - Betting team must meet or exceed their bid in points
 * - Opposition threshold is 304 - bid + 1
 * - Stone given (win) or received (loss) varies by bid level
 * - Caps modifiers apply on top of normal scoring
 * - PCC: 5 stone win or loss
 */

const { SCORING_TABLE, PCC_SCORING, TOTAL_POINTS, INITIAL_STONE, PHASE, SEATS } = require("./constants");
const { getTeamForSeat, getPartnerSeat, getDealOrder, getNextSeat, getCutterSeat, createPack, minimalShuffle } = require("./deck");
const { generateAllPlayerViews } = require("./views");
const { getFirestore } = require("firebase-admin/firestore");

const db = () => getFirestore();

/**
 * Calculate the result of a completed game.
 *
 * @param {Object} game - The full game state
 * @returns {Object} Result with reason, stoneExchanged, stoneDirection, winnerTeam
 */
function calculateResult(game) {
  const bidding = game.bidding;
  const play = game.play;
  const trump = game.trump;

  const trumperSeat = trump.trumperSeat;
  const trumperTeam = getTeamForSeat(trumperSeat);
  const oppositionTeam = trumperTeam === "teamA" ? "teamB" : "teamA";

  // Check for caps situations first
  if (play.capsCall) {
    // Caps result already determined in caps.js
    return game.result;
  }

  // Check for late caps (player was obligated but never called)
  for (const seat of SEATS) {
    if (play.capsObligations && play.capsObligations[seat]) {
      const obligation = play.capsObligations[seat];
      const obligatedTeam = getTeamForSeat(seat);

      // Check if the team actually won all rounds from the obligation point onward
      const allWon = play.completedRounds
        .filter((r) => r.roundNumber >= obligation.obligatedAtRound)
        .every((r) => getTeamForSeat(r.winner) === obligatedTeam);

      if (allWon) {
        // Late caps — penalty
        const bid = bidding.isPCC ? "pcc" : bidding.highestBid;
        const scoring = bidding.isPCC ? PCC_SCORING : SCORING_TABLE[bid];
        const normalLoss = scoring ? scoring.loss : 2;

        return {
          reason: "caps_late",
          stoneExchanged: normalLoss + 1,
          stoneDirection: "receive",
          winnerTeam: obligatedTeam === "teamA" ? "teamB" : "teamA",
          capsBy: seat,
          lateDetected: true,
          description: `Late Caps detected for ${seat}. ${normalLoss + 1} stone penalty.`,
        };
      }
    }
  }

  // PCC scoring
  if (bidding.isPCC) {
    const allWon = play.completedRounds.every(
      (r) => getTeamForSeat(r.winner) === trumperTeam
    );

    if (allWon) {
      return {
        reason: "pcc_won",
        stoneExchanged: PCC_SCORING.win,
        stoneDirection: "give",
        winnerTeam: trumperTeam,
        description: `PCC successful! ${PCC_SCORING.win} stone given.`,
      };
    } else {
      return {
        reason: "pcc_lost",
        stoneExchanged: PCC_SCORING.loss,
        stoneDirection: "receive",
        winnerTeam: oppositionTeam,
        description: `PCC failed. ${PCC_SCORING.loss} stone received.`,
      };
    }
  }

  // Normal scoring
  const bid = bidding.highestBid;
  const scoring = SCORING_TABLE[bid];
  if (!scoring) {
    return {
      reason: "error",
      stoneExchanged: 0,
      stoneDirection: "none",
      winnerTeam: null,
      description: `Unknown bid value: ${bid}`,
    };
  }

  const trumperPoints = play.pointsWon[trumperTeam];
  const oppositionPoints = play.pointsWon[oppositionTeam];

  if (trumperPoints >= bid) {
    // Betting team met their bid
    return {
      reason: "bid_met",
      stoneExchanged: scoring.win,
      stoneDirection: "give",
      winnerTeam: trumperTeam,
      trumperPoints,
      oppositionPoints,
      bid,
      description: `Bid of ${scoring.name} met with ${trumperPoints} points. ${scoring.win} stone given.`,
    };
  } else {
    // Betting team failed
    return {
      reason: "bid_failed",
      stoneExchanged: scoring.loss,
      stoneDirection: "receive",
      winnerTeam: oppositionTeam,
      trumperPoints,
      oppositionPoints,
      bid,
      description: `Bid of ${scoring.name} failed with ${trumperPoints} points (needed ${bid}). ${scoring.loss} stone received.`,
    };
  }
}

/**
 * Start a new game within the same match (next dealer).
 *
 * @param {Object} data - { code: string }
 * @param {Object} context - Firebase callable context
 */
async function nextGame(data, context) {
  const uid = context.auth?.uid;
  if (!uid) throw new Error("Authentication required.");

  const code = (data.code || "").toUpperCase().trim();
  const lobbyRef = db().collection("lobbies").doc(code);

  const result = await db().runTransaction(async (transaction) => {
    const lobbyDoc = await transaction.get(lobbyRef);
    if (!lobbyDoc.exists) throw new Error("Game not found.");
    const lobby = lobbyDoc.data();

    const gameRef = lobbyRef.collection("games").doc(lobby.gameId);
    const gameDoc = await transaction.get(gameRef);
    if (!gameDoc.exists) throw new Error("Game not found.");
    const prevGame = gameDoc.data();

    if (prevGame.phase !== PHASE.COMPLETE) throw new Error("Current game is not complete.");

    // Check if match is over
    const stone = prevGame.stone || { teamA: INITIAL_STONE, teamB: INITIAL_STONE };
    if (stone.teamA <= 0) {
      return { matchComplete: true, winner: "teamA" };
    }
    if (stone.teamB <= 0) {
      return { matchComplete: true, winner: "teamB" };
    }

    // Create new game with next dealer
    const nextDealer = getNextSeat(prevGame.dealerSeat);
    const deck = minimalShuffle(createPack());
    const cutterSeat = getCutterSeat(nextDealer);

    const newGameState = {
      gameNumber: (prevGame.gameNumber || 1) + 1,
      dealerSeat: nextDealer,
      phase: PHASE.DEALING_4,
      stone: stone,
      deck: deck,
      hands: { north: [], east: [], south: [], west: [] },
      trump: {
        trumperSeat: null,
        trumpSuit: null,
        trumpCard: null,
        isRevealed: false,
        isOpen: false,
        trumpCardInHand: false,
      },
      bidding: null,
      play: null,
      reshuffleCount: 0,
      consecutiveReshuffles: 0,
      result: null,
      cutting: {
        cutterSeat,
        resolved: false,
        didCut: null,
      },
      pccPartnerOut: null,
    };

    const newGameRef = lobbyRef.collection("games").doc();
    transaction.set(newGameRef, newGameState);
    transaction.update(lobbyRef, { gameId: newGameRef.id });

    const views = generateAllPlayerViews(newGameState, lobby);
    for (const seat of SEATS) {
      if (views[seat]) {
        const viewRef = newGameRef.collection("playerViews").doc(seat);
        transaction.set(viewRef, views[seat]);
      }
    }

    return { matchComplete: false, gameId: newGameRef.id };
  });

  return result;
}

module.exports = {
  calculateResult,
  nextGame,
};

File: game304/scoring.py
"""Scoring and stone exchange for the 304 card game.

Calculates game results and applies stone changes at the end of each
game. Handles normal scoring, PCC, caps modifiers, and special
endings (spoilt trumps, absolute hand).

Scoring rules:
- Betting team must meet or exceed their bid in points.
- Opposition threshold is ``304 - bid + 1``.
- Stone given (win) or received (loss) varies by bid level.
- Caps modifiers apply on top of normal scoring.
- PCC: 5 stone win or 5 stone loss.
"""

from __future__ import annotations

from game304.constants import (
    PCC_BID_VALUE,
    PCC_SCORING,
    SCORING_TABLE,
    WRONG_CAPS_PENALTY,
)
from game304.seating import team_of
from game304.state import GameResult, GameState
from game304.types import Seat, Team


def _other_team(team: Team) -> Team:
    """Return the opposing team."""
    return Team.TEAM_B if team == Team.TEAM_A else Team.TEAM_A


def calculate_result(state: GameState) -> GameResult:
    """Calculate the result of a completed game.

    Checks in priority order:
    1. Caps call results (already resolved during play).
    2. Late caps detection (player was obligated but never called).
    3. PCC scoring (all-or-nothing).
    4. Normal scoring (bid met or failed).

    Args:
        state: The completed game state.

    Returns:
        A ``GameResult`` describing the outcome.
    """
    bidding = state.bidding
    play = state.play
    trump = state.trump

    trumper_seat = trump.trumper_seat
    trumper_team = team_of(trumper_seat)
    opposition_team = _other_team(trumper_team)

    # 1. If caps was already called and resolved, return existing result
    if play.caps_call is not None and state.result is not None:
        return state.result

    # 2. PCC scoring — caps and external caps mechanics do not apply.
    #    Late-caps detection must be skipped before checking obligations.
    if bidding.is_pcc:
        all_won = all(
            team_of(r.winner) == trumper_team for r in play.completed_rounds
        )
        if all_won:
            return GameResult(
                reason="pcc_won",
                stone_exchanged=PCC_SCORING.win,
                stone_direction="give",
                winner_team=trumper_team,
                description=(
                    f"PCC successful! {PCC_SCORING.win} stone given."
                ),
            )
        else:
            return GameResult(
                reason="pcc_lost",
                stone_exchanged=PCC_SCORING.loss,
                stone_direction="receive",
                winner_team=opposition_team,
                description=(
                    f"PCC failed. {PCC_SCORING.loss} stone received."
                ),
            )

    # 3. Check for late caps (player was obligated but never called caps).
    #    A player obligated mid-game whose team did go on to win every
    #    remaining round is treated as Late Caps: the betting team's win
    #    flips to a loss + 1 stone (or the opposition's win to a loss + 1
    #    in the external case).
    if play.caps_obligations:
        for seat_key, obligation in play.caps_obligations.items():
            obligated_team = team_of(seat_key)
            rounds_after = [
                r
                for r in play.completed_rounds
                if r.round_number >= obligation.obligated_at_round
            ]
            all_won = bool(rounds_after) and all(
                team_of(r.winner) == obligated_team for r in rounds_after
            )
            # Also require: the obligated team won every round of the
            # game (the precondition for caps in the first place).
            won_everything = all(
                team_of(r.winner) == obligated_team
                for r in play.completed_rounds
            )
            if all_won and won_everything:
                scoring = _get_scoring(bidding)
                return GameResult(
                    reason="caps_late",
                    stone_exchanged=scoring.loss + 1,
                    stone_direction="receive",
                    winner_team=_other_team(obligated_team),
                    caps_by=seat_key,
                    description=(
                        f"Late Caps detected for {seat_key.value}. "
                        f"{scoring.loss + 1} stone penalty."
                    ),
                )

    # 4. Normal scoring
    bid = bidding.highest_bid
    scoring = SCORING_TABLE.get(bid)
    if scoring is None:
        return GameResult(
            reason="error",
            stone_exchanged=0,
            stone_direction="none",
            winner_team=None,
            description=f"Unknown bid value: {bid}",
        )

    trumper_points = play.points_won[trumper_team]
    opposition_points = play.points_won[opposition_team]

    if trumper_points >= bid:
        return GameResult(
            reason="bid_met",
            stone_exchanged=scoring.win,
            stone_direction="give",
            winner_team=trumper_team,
            trumper_points=trumper_points,
            opposition_points=opposition_points,
            bid=bid,
            description=(
                f"Bid of {scoring.name} met with {trumper_points} points. "
                f"{scoring.win} stone given."
            ),
        )
    else:
        return GameResult(
            reason="bid_failed",
            stone_exchanged=scoring.loss,
            stone_direction="receive",
            winner_team=opposition_team,
            trumper_points=trumper_points,
            opposition_points=opposition_points,
            bid=bid,
            description=(
                f"Bid of {scoring.name} failed with {trumper_points} "
                f"points (needed {bid}). {scoring.loss} stone received."
            ),
        )


def apply_stone_changes(
    stone: dict[Team, int],
    result: GameResult,
    trumper_team: Team,
) -> None:
    """Apply stone changes based on the game result.

    Modifies the stone dict in place.

    Args:
        stone: Current stone counts per team (mutated in place).
        result: The game result.
        trumper_team: The team that held the bid.
    """
    if result.stone_direction == "none":
        return

    if result.stone_direction == "give":
        # Betting team won — they give stone (subtract from count)
        stone[trumper_team] = max(0, stone[trumper_team] - result.stone_exchanged)
    elif result.stone_direction == "receive":
        # Betting team lost — they receive stone (add to count)
        stone[trumper_team] += result.stone_exchanged


def calculate_caps_result(
    state: GameState,
    seat: Seat,
    is_valid: bool,
    is_external: bool,
) -> GameResult:
    """Calculate the result of a caps call.

    Handles correct, late, and wrong/early caps for both regular
    and external (opposition) caps calls.

    Args:
        state: The game state.
        seat: The seat that called caps.
        is_valid: Whether the caps call is valid (can guarantee all rounds).
        is_external: Whether this is external caps (from the opposition).

    Returns:
        A ``GameResult`` describing the caps outcome.
    """
    my_team = team_of(seat)
    trumper_team = team_of(state.trump.trumper_seat)

    if not is_valid:
        # Wrong/Early Caps — 5 stone penalty to calling team
        return GameResult(
            reason="caps_wrong",
            stone_exchanged=WRONG_CAPS_PENALTY,
            stone_direction="receive",
            winner_team=_other_team(my_team),
            caps_by=seat,
            description=(
                f"Wrong/Early Caps by {seat.value}. "
                f"{WRONG_CAPS_PENALTY} stone penalty."
            ),
        )

    # Valid caps — check timing
    is_late = False
    play = state.play
    if play is not None:
        from game304.caps import is_caps_late

        is_late = is_caps_late(state, seat)

    is_before_round_7 = play is not None and play.round_number < 7
    scoring = _get_scoring(state.bidding)

    if is_late:
        # Late caps — loss + 1 stone penalty
        normal_loss = scoring.loss
        return GameResult(
            reason="caps_late",
            stone_exchanged=normal_loss + 1,
            stone_direction="receive",
            winner_team=_other_team(my_team),
            caps_by=seat,
            description=(
                f"Late Caps by {seat.value}. "
                f"{normal_loss + 1} stone penalty."
            ),
        )

    if is_before_round_7:
        # Correct caps before Round 7 — bonus stone
        if is_external:
            normal_loss = scoring.loss
            return GameResult(
                reason="external_caps",
                stone_exchanged=normal_loss + 1,
                stone_direction="receive",
                winner_team=my_team,
                caps_by=seat,
                description=(
                    f"External Caps (correct, before Round 7). "
                    f"Betting team receives {normal_loss + 1} stone."
                ),
            )
        else:
            normal_win = scoring.win
            return GameResult(
                reason="caps_correct",
                stone_exchanged=normal_win + 1,
                stone_direction="give",
                winner_team=my_team,
                caps_by=seat,
                description=(
                    f"Caps correct (before Round 7). "
                    f"Betting team gives {normal_win + 1} stone."
                ),
            )
    else:
        # Correct caps after Round 7 — no bonus, normal scoring
        if is_external:
            normal_loss = scoring.loss
            return GameResult(
                reason="external_caps",
                stone_exchanged=normal_loss,
                stone_direction="receive",
                winner_team=my_team,
                caps_by=seat,
                description=(
                    f"External Caps (correct, after Round 7). "
                    f"Normal loss applies."
                ),
            )
        else:
            normal_win = scoring.win
            return GameResult(
                reason="caps_correct",
                stone_exchanged=normal_win,
                stone_direction="give",
                winner_team=my_team,
                caps_by=seat,
                description=(
                    f"Caps correct (after Round 7). Normal win applies."
                ),
            )


def _get_scoring(bidding):
    """Get the scoring entry for the current bid."""
    if bidding.is_pcc:
        return PCC_SCORING
    return SCORING_TABLE.get(bidding.highest_bid, PCC_SCORING)
